### ISSUE TITLE
Legion grid is sometimes inaccurate

### ISSUE THREAD

**Issue by kendonB** [synkarius/caster#172]
I think this is because I switch between two monitors. Is there a way to reset the parameters that go into legion?

![legion](https://cloud.githubusercontent.com/assets/5327505/22630734/e416ada2-ebb3-11e6-8393-5b027570be0c.png)

**Comment by chilimangoes** [comment]
Man, this is weird. @imalloutofbubblegum [reported the same issue on Gitter](https://gitter.im/synkarius/caster?at=5850b462aeb490080487baf0), but for me it works fine on multiple monitors both at home and work. There used to be [an issue](https://github.com/synkarius/caster/issues/116) where it would always use a screenshot of the primary monitor for the Legion command, regardless of which monitor you displayed the grid on, but that should have been fixed by [this commit](https://github.com/synkarius/caster/commit/8eb74eb43fbd05d742cb19789f37c32ea4b7382b).

Some questions and things to look at that might help track down the issue:

1) What version of Windows are you running?
2) Is this sporadic or can you reproduce it reliably? Does it always happen on one monitor but not the other (e.g. only on your secondary monitor)?
3) When you look at the code you're running, do you see the changes from the fix noted above?
4) Are you comfortable enough with Python to try testing the `grab_screen` method with different `bbox` parameters to see if what is captured by the method actually corresponds to what is being requested of it? If not, I should be able to work up some code for you to run, but it'll probably take me a couple days.

**Comment by kendonB** [comment]
1) I'm running Windows 10
2) It isn't sporadic exactly, but I can't reproduce it either. If it's happening, it seems to always happen until I restart the computer. Even then it can happen instantly as well.
3) I can see all the changes in that commit, yes (I downloaded caster in the last couple of weeks).
4) I worked it out. I tried these and they seemed to be correct:
```

>>> grab_screen([0, 0, 400, 400]).show()
>>>
>>> grab_screen([0, 0, 2000, 2000]).show()
```

**Comment by chilimangoes** [comment]
Hmmm, curiouser and curiouser. It sounds like I might need to add some debugging code in there to output some additional info to the NatLink window to give us an idea of exactly where things are getting mixed up.

Regarding your answers to #2:

2a) Have you noticed it happening only on your secondary monitor, or does it happen on your primary as well? And if it happens to both, does it happen to both simultaneously, or always just one or the other.

2b) Does restarting Dragon work (maybe try restarting Dragon a couple times) or do you actually have to restart your computer?

2c) Any sense of what the frequency of this happening is, like 25% of the time, 50%, or more?

Another question, just to check my assumptions, when you say you switch between two monitors, am I correct that you mean you use two monitors simultaneously and span your desktop over both of them? I'm asking because actually live switching monitor configurations isn't something I tested for.

**Comment by kendonB** [comment]
2a) I've noticed on each monitor separately, but at different times. Only now I've tested with both at the same time. Currently, it is working fine on my big monitor, and I'm seeing the error on my laptop monitor. I tried the grab screen method and it seemed to be working fine, but I wasn't all that careful. 

2b) Restarting Dragon doesn't seem to work. However, restarting my computer **also** doesn't seem to work, at the moment. Currently, I am unable to get it to work on my laptop screen at all.

2c) Hard to say; right now, it's 100% on my laptop monitor and 0% on my big screen. It could be that it's somehow constantly switching between the two.

Regarding the last question, I almost never use both monitors simultaneously; just one or the other (big screen while at work and laptop screen while at home)

**Comment by kendonB** [comment]
I believe I've solved it. Windows will resize the text on one monitor so that everything is approximately the same size. It seems to even do this persistently after the second monitor is disconnected. In Display Settings, there is an option to "Change the size of text, apps, and other items", and I reset this to 100%.

Now the only problem is Legion not staying on top.

**Comment by kendonB** [comment]
I've come across another related problem. Legion seems to stop at a bounding box that is smaller than the total screen. This is now on my big monitor. It looks like it stops at approximately the size of the smaller screen.

Legion is behaving correctly on the small motor, currently.

![legion_example](https://cloud.githubusercontent.com/assets/5327505/22752621/110e7f74-edee-11e6-8209-85e7ac8c71a2.png)

**Comment by kendonB** [comment]
Now, the problem is solved by rebooting Dragon. Confirmed it is working on both monitors correctly.

**Comment by chilimangoes** [comment]
Wow, DPI settings. That actually makes perfect sense now, and I can reproduce it too by scaling my DPI settings above 100%. I'm gonna need to do some thinking about how this problem can be solved without telling people they need to run their DPI settings at 100%. To be continued, I guess.

**Comment by avigod** [comment]
Setting DPI awareness might help? 
https://stackoverflow.com/a/32541666

**Comment by gerrish** [comment]
I have fixed this for my monitor, but with a work around. The positions of the text seem to be correctly calculated , however the DPI scaling setting seems to lead to the Legion .grid getting wrong startup dimensions. Simply multiplying the dx,dy values by the scaling factor seems to fix this. However I can't seem to find an automatic way of getting to that scaling factor. I have set it to a constant value but for multiple monitors or a variable scaling factor this is suboptimal.

**Comment by chilimangoes** [comment]
This should be fixed now as part of pull request #223. If anyone has a test case where this still fails, please comment and we will reopen the issue. @kendonB Could you check the referenced pull request to see if it solves the problem you were seeing?
